File: Problems/0300_0399/0304_Range_Sum_Query_2D-Immutable/Project_Java/NumMatrix.java
class NumMatrix {
    int[][] sum;

    public NumMatrix(int[][] arg_matrix) {
        sum = new int[arg_matrix.length][];
        for (int i = 0; i < sum.length; i++) {
            int tmp = 0;
            sum[i] = new int[arg_matrix[i].length];
            for (int j = 0; j < arg_matrix[i].length; j++) {
                int tmp2 = 0;
                if (i != 0)
                    tmp2 = sum[i - 1][j];
                sum[i][j] = tmp + arg_matrix[i][j] + tmp2;
                tmp += arg_matrix[i][j];
            }
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        int blk1 = 0;

        if (row1 != 0)
            blk1 = sum[row1-1][col2];

        int blk2 = 0;
        if (col1 != 0)
            blk2 = sum[row2][col1-1];

        int blk3 = 0;
        if (row1 != 0 && col1 != 0)
            blk3 = sum[row1-1][col1-1];

        int blk4 = sum[row2][col2];

        return blk4 - blk1 - blk2 + blk3;
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */

File: Problems/0300_0399/0304_Range_Sum_Query_2D-Immutable/Project_Go/solution/nummatrix/nummatrix.go
package nummatrix

import "fmt"

type NumMatrix struct {
	sum [][]int
}

func Constructor(matrix [][]int) NumMatrix {
	nm := new(NumMatrix)
	h := len(matrix)
	nm.sum = make([][]int, len(matrix))

	if len(matrix) < 1 {
		return *nm
	}

	if len(matrix[0]) < 1 {
		return *nm
	}

	w := len(matrix[0])
	for i := 0; i < h; i++ {
		nm.sum[i] = make([]int, w)
	}

	for i := 0; i < h; i++ {
		tmp := 0
		for j := 0; j < w; j++ {
			tmp2 := 0
			if i != 0 {
				tmp2 = nm.sum[i-1][j]
			}
			nm.sum[i][j] = tmp + matrix[i][j] + tmp2
			tmp += matrix[i][j]
		}
	}

	return *nm
}

func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
	blk1 := 0
	if row1 != 0 {
		blk1 = this.sum[row1-1][col2]
	}
	blk2 := 0
	if col1 != 0 {
		blk2 = this.sum[row2][col1-1]
	}
	blk3 := 0
	if row1 != 0 && col1 != 0 {
		blk3 = this.sum[row1-1][col1-1]
	}
	blk4 := this.sum[row2][col2]

	return blk4 - blk1 - blk2 + blk3
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * obj := Constructor(matrix);
 * param_1 := obj.SumRegion(row1,col1,row2,col2);
 */

func (this *NumMatrix) PrintMatrix() {
	fmt.Printf("[\n")
	for i := 0; i < len(this.sum); i++ {
		fmt.Printf(" [")
		for j := 0; j < len(this.sum[i]); j++ {
			fmt.Printf(" %d", this.sum[i][j])
		}
		fmt.Printf(" ]\n")
	}
	fmt.Printf("]\n")
}
